File: packages/external-idp/android/src/main/java/com/pingidentity/rnexternalidp/RNPingExternalIdpCommon.kt
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */
package com.pingidentity.rnexternalidp

import android.content.ActivityNotFoundException
import android.net.Uri
import androidx.annotation.VisibleForTesting
import androidx.core.net.toUri
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import com.facebook.react.bridge.ReadableType
import com.pingidentity.android.ContextProvider
import com.pingidentity.idp.IdpCanceledException
import com.pingidentity.idp.UnsupportedIdPException
import com.pingidentity.idp.journey.IdpCallback
import com.pingidentity.idp.journey.SelectIdpCallback
import com.pingidentity.logger.Logger
import com.pingidentity.rncore.CoreRuntime
import com.pingidentity.rncore.error.ErrorType
import com.pingidentity.rncore.error.GenericError
import com.pingidentity.rncore.error.mapThrowableToGenericError
import com.pingidentity.rncore.error.reject
import com.pingidentity.rncore.logger.LoggerHandleContract
import com.pingidentity.rncore.utils.JsonBridgeMapper
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

/**
 * Shared implementation for External IdP operations on Android.
 */
object RNPingExternalIdpCommon {
  private const val LOGGER_ID_KEY = "loggerId"
  private const val REDIRECT_URI_KEY = "redirectUri"
  private const val INDEX_KEY = "index"

  /** Coroutine scope for executing External IdP operations asynchronously on the IO dispatcher. */
  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.IO)

  /**
   * Per-call External IdP runtime configuration.
   */
  private data class CallConfig(val loggerId: String?, val redirectUri: String)

  /**
   * Activity availability provider used by runtime checks and overridable in tests.
   */
  @VisibleForTesting
  @JvmSynthetic
  internal var foregroundActivityProvider: () -> Boolean = {
    runCatching { ContextProvider.currentActivity }
      .getOrNull() != null
  }

  /**
   * Configure application context required by Ping native SDKs.
   */
  @JvmStatic
  fun configure(reactContext: ReactApplicationContext) {
    ContextProvider.init(reactContext.applicationContext)
  }

  /**
   * Launches the external IdP authorization flow for an active Journey IdpCallback.
   *
   * @param journeyId Native Journey instance id.
   * @param options Per-call options payload (index).
   * @param config Per-call configuration payload (loggerId, redirectUri).
   * @param promise React Native promise resolved with ExternalIdpResult or rejected on error.
   */
  @JvmStatic
  fun authorizeForJourney(
    journeyId: String,
    options: ReadableMap,
    config: ReadableMap,
    promise: Promise
  ) {
    val callConfig = parseCallConfig(config)
    val logger = resolveLoggerFromCore(callConfig.loggerId)
    if (journeyId.isBlank()) {
      logger?.w("External IdP authorizeForJourney rejected because journey id was empty", null)
      rejectWithError(
        promise = promise,
        code = ExternalIdpErrorCodes.CALLBACK_NOT_FOUND,
        message = "Journey id must not be empty for external IdP authorization.",
        type = ErrorType.ARGUMENT_ERROR
      )
      return
    }
    if (!hasForegroundActivity()) {
      logger?.w(
        "External IdP authorizeForJourney rejected because no foreground activity is available",
        null
      )
      rejectWithError(
        promise = promise,
        code = ExternalIdpErrorCodes.ACTIVITY_UNAVAILABLE,
        message = "No foreground activity available for external IdP authorization."
      )
      return
    }

    scope.launch {
      try {
        val index = parseCallbackIndex(options)
        logger?.i("External IdP authorizeForJourney requested for callback index $index")
        val callback = resolveIdpCallback(journeyId, index)
        if (callback == null) {
          logger?.w("External IdP authorizeForJourney callback not found at index $index", null)
          rejectWithError(
            promise = promise,
            code = ExternalIdpErrorCodes.CALLBACK_NOT_FOUND,
            message = "No active IdP callback found for journey $journeyId at index $index.",
            type = ErrorType.STATE_ERROR
          )
          return@launch
        }
        val parsedRedirectUri = try {
          parseRedirectUri(callConfig.redirectUri, callback.provider)
        } catch (e: IllegalArgumentException) {
          logger?.w("External IdP authorizeForJourney rejected because redirect URI was invalid", e)
          rejectWithError(
            promise = promise,
            code = ExternalIdpErrorCodes.CONFIG_ERROR,
            message = e.localizedMessage ?: "Redirect URI is invalid.",
            type = ErrorType.ARGUMENT_ERROR,
            throwable = e
          )
          return@launch
        }

        // TODO: Add browser fallback here when the Android native SDK supports it for AIC Journey flows.
        val result = withContext(Dispatchers.Main) {
          parsedRedirectUri?.let { callback.authorize(it) } ?: callback.authorize()
        }

        result.fold(
          onSuccess = { idpResult ->
            logger?.d("External IdP authorizeForJourney succeeded")
            val resultJson = buildJsonObject {
              put("token", JsonPrimitive(idpResult.token))
              put(
                "additionalParameters",
                buildJsonObject {
                  idpResult.additionalParameters.forEach { (k, v) -> put(k, JsonPrimitive(v)) }
                }
              )
            }
            promise.resolve(JsonBridgeMapper.encodeJsonObject(resultJson))
          },
          onFailure = { error ->
            val (code, message) = discriminateIdpError(error)
            logger?.e("External IdP authorizeForJourney failed with code $code", error)
            rejectWithError(
              promise = promise,
              code = code,
              message = message,
              throwable = error
            )
          }
        )
      } catch (e: Throwable) {
        when (e) {
          is ClassNotFoundException, is NoClassDefFoundError -> {
            logger?.e("External IdP authorizeForJourney failed because provider SDK is unavailable", e)
            rejectWithError(
              promise = promise,
              code = ExternalIdpErrorCodes.UNSUPPORTED_PROVIDER,
              message = "Native provider SDK is not available: ${e.message}",
              throwable = e
            )
          }
          else -> {
            val (code, message) = discriminateIdpError(e)
            logger?.e("External IdP authorizeForJourney failed with code $code", e)
            rejectWithError(
              promise = promise,
              code = code,
              message = message,
              throwable = e
            )
          }
        }
      }
    }
  }

  /**
   * Mutates the native SelectIdpCallback state for an active Journey.
   *
   * Journey's JourneyCallbackValueApplier blocks SelectIdPCallback and throws
   * MissingIntegrationException when it appears in a journey.next() payload. This bridge
   * call must be made before journey.next() so the callback state is already populated natively.
   *
   * @param journeyId Native Journey instance id.
   * @param provider The provider string selected by the user.
   * @param options Per-call options payload (index).
   * @param config Per-call configuration payload (loggerId).
   * @param promise React Native promise resolved on success or rejected on error.
   */
  @JvmStatic
  fun setSelectedProvider(
    journeyId: String,
    provider: String,
    options: ReadableMap,
    config: ReadableMap,
    promise: Promise
  ) {
    val logger = resolveCallLogger(config)
    if (journeyId.isBlank()) {
      logger?.w("External IdP select provider rejected because journey id was empty", null)
      rejectWithError(
        promise = promise,
        code = ExternalIdpErrorCodes.CALLBACK_NOT_FOUND,
        message = "Journey id must not be empty for setting selected provider.",
        type = ErrorType.ARGUMENT_ERROR
      )
      return
    }
    val selectedProvider = provider.trim()
    if (selectedProvider.isBlank()) {
      logger?.w("External IdP select provider rejected because provider was empty", null)
      rejectWithError(
        promise = promise,
        code = ExternalIdpErrorCodes.CONFIG_ERROR,
        message = "Provider must not be empty for external IdP selection.",
        type = ErrorType.ARGUMENT_ERROR
      )
      return
    }

    scope.launch {
      try {
        val index = parseCallbackIndex(options)
        logger?.i("External IdP select provider requested for callback index $index")
        val callback = resolveSelectIdpCallback(journeyId, index)
        if (callback == null) {
          logger?.w("External IdP select provider callback not found at index $index", null)
          rejectWithError(
            promise = promise,
            code = ExternalIdpErrorCodes.CALLBACK_NOT_FOUND,
            message = "No active SelectIdp callback found for journey $journeyId at index $index.",
            type = ErrorType.STATE_ERROR
          )
          return@launch
        }

        callback.value = selectedProvider
        logger?.d("External IdP select provider succeeded")
        promise.resolve(null)
      } catch (e: Throwable) {
        logger?.e("External IdP select provider failed", e)
        rejectWithError(
          promise = promise,
          code = ExternalIdpErrorCodes.CONFIG_ERROR,
          message = e.localizedMessage ?: "Failed to set selected provider.",
          throwable = e
        )
      }
    }
  }

  /**
   * Rejects a promise with the shared External IdP error contract.
   */
  private fun rejectWithError(
    promise: Promise,
    code: String,
    message: String,
    type: ErrorType? = null,
    throwable: Throwable? = null
  ) {
    val mapped = throwable?.let { mapThrowableToGenericError(it, code) }
    val resolvedType = type ?: mapped?.type ?: ErrorType.AUTH_ERROR
    val resolvedMessage = message.ifBlank { mapped?.message ?: "Unknown error" }
    val error = GenericError(
      type = resolvedType,
      error = code,
      message = resolvedMessage
    )
    promise.reject(error, throwable)
  }

  /**
   * Returns whether a foreground Android activity is available for UI-based authorization.
   */
  private fun hasForegroundActivity(): Boolean = foregroundActivityProvider()

  /**
   * Parses and validates the optional redirect URI passed to the native External IdP SDK.
   *
   * @param redirectUri Optional app return URI configured for Auth Tab-capable devices.
   * @param provider IdP provider resolved from the active Journey callback.
   * @return Parsed Android URI, or null to delegate to native defaults.
   * @throws IllegalArgumentException when an Apple redirect URI is missing a scheme.
   */
  private fun parseRedirectUri(redirectUri: String, provider: String): Uri? {
    val trimmed = redirectUri.trim()
    if (trimmed.isEmpty()) return null
    val uri = trimmed.toUri()
    require(!isAppleProvider(provider) || !uri.scheme.isNullOrBlank()) {
      "Redirect URI must include a URI scheme for external IdP authorization."
    }
    return uri
  }

  /**
   * Returns whether the Journey IdP provider requires a redirect URI override.
   *
   * @param provider IdP provider resolved from the active Journey callback.
   * @return True when the provider is Apple, false otherwise.
   */
  private fun isAppleProvider(provider: String): Boolean {
    return provider.contains("apple", ignoreCase = true)
  }

  /**
   * Resolves the configured per-call logger.
   *
   * @param config Per-call configuration payload.
   * @return Native logger resolved from Core, or null when no valid logger is configured.
   */
  private fun resolveCallLogger(config: ReadableMap): Logger? {
    return resolveLoggerFromCore(parseCallConfig(config).loggerId)
  }

  /**
   * Parses per-call configuration payload.
   *
   * @param config Per-call configuration payload.
   * @return Parsed External IdP runtime configuration.
   */
  private fun parseCallConfig(config: ReadableMap): CallConfig {
    val loggerId = if (
      config.hasKey(LOGGER_ID_KEY) && config.getType(LOGGER_ID_KEY) == ReadableType.String
    ) {
      config.getString(LOGGER_ID_KEY)?.trim()?.takeIf { it.isNotEmpty() }
    } else {
      null
    }
    val redirectUri = if (
      config.hasKey(REDIRECT_URI_KEY) && config.getType(REDIRECT_URI_KEY) == ReadableType.String
    ) {
      config.getString(REDIRECT_URI_KEY)?.trim() ?: ""
    } else {
      ""
    }
    return CallConfig(loggerId = loggerId, redirectUri = redirectUri)
  }

  /**
   * Resolve a native logger from the shared Core logger registry.
   *
   * @param id Logger handle identifier from JS.
   * @return Native logger instance, or null when missing/invalid.
   */
  private fun resolveLoggerFromCore(id: String?): Logger? {
    if (id.isNullOrBlank()) return null
    val handle = CoreRuntime.loggerRegistry.resolve(id) as? LoggerHandleContract ?: return null
    return handle.nativeLogger as? Logger
  }

  /**
   * Resolves an IdpCallback from core-exposed Journey callbacks by journeyId and type index.
   */
  private suspend fun resolveIdpCallback(journeyId: String, index: Int): IdpCallback? {
    val callbacks = CoreRuntime.resolveJourneyCallbacks(journeyId) ?: return null
    return callbacks.filterIsInstance<IdpCallback>().getOrNull(index)
  }

  /**
   * Resolves a SelectIdpCallback from core-exposed Journey callbacks by journeyId and type index.
   */
  private suspend fun resolveSelectIdpCallback(journeyId: String, index: Int): SelectIdpCallback? {
    val callbacks = CoreRuntime.resolveJourneyCallbacks(journeyId) ?: return null
    return callbacks.filterIsInstance<SelectIdpCallback>().getOrNull(index)
  }

  /**
   * Parses the callback type index from options, defaulting to 0.
   */
  private fun parseCallbackIndex(options: ReadableMap?): Int {
    if (options == null || !options.hasKey(INDEX_KEY) || options.isNull(INDEX_KEY)) return 0
    return when (options.getType(INDEX_KEY)) {
      ReadableType.Number -> options.getDouble(INDEX_KEY).toInt()
      ReadableType.String -> options.getString(INDEX_KEY)?.toIntOrNull() ?: 0
      else -> 0
    }
  }

  /**
   * Maps an IdP authorize failure to its error code and message.
   */
  private fun discriminateIdpError(error: Throwable): Pair<String, String> = when {
    error is IdpCanceledException || isCredentialCancellation(error) ->
      ExternalIdpErrorCodes.CANCELLED to
        (error.localizedMessage ?: "User cancelled external IdP authorization.")
    error is UnsupportedIdPException ->
      ExternalIdpErrorCodes.UNSUPPORTED_PROVIDER to
        (error.localizedMessage ?: "Unsupported IdP provider.")
    error is IllegalArgumentException ->
      ExternalIdpErrorCodes.CONFIG_ERROR to
        (error.localizedMessage ?: "External IdP provider configuration is invalid.")
    error is ActivityNotFoundException ->
      ExternalIdpErrorCodes.ACTIVITY_UNAVAILABLE to
        (error.localizedMessage ?: "No activity available for external IdP authorization.")
    else ->
      ExternalIdpErrorCodes.AUTHORIZE_ERROR to
        (error.localizedMessage ?: "External IdP authorization failed.")
  }

  /**
   * Returns whether the throwable chain represents an Android Credential Manager cancellation.
   *
   * @param error Throwable emitted by the native IdP SDK.
   * @return True when AndroidX Credential Manager reported user cancellation or reauth cancellation.
   */
  private fun isCredentialCancellation(error: Throwable): Boolean {
    var current: Throwable? = error
    while (current != null) {
      val className = current::class.qualifiedName
      val simpleName = current::class.simpleName
      if (
        className == "androidx.credentials.exceptions.GetCredentialCancellationException" ||
        simpleName == "GetCredentialCancellationException"
      ) {
        return true
      }
      current = current.cause
    }
    return false
  }
}

File: packages/external-idp/ios/RNPingExternalIdpCommon.swift
/*
 * Copyright (c) 2026 Ping Identity Corporation. All rights reserved.
 *
 * This software may be modified and distributed under the terms
 * of the MIT license. See the LICENSE file for details.
 */

import Foundation
import PingExternalIdP
import PingLogger
import React
import RNPingCore

/// Per-call External IdP runtime configuration resolved from the JS bridge payload.
struct ExternalIdpCallConfig {
  /// Optional logger handle identifier forwarded from JS.
  let loggerId: String?

  /// Optional app return URI override forwarded from JS.
  let redirectUri: String
}

/// IdP handler decorator that overrides the redirect URI before native authorization starts.
@MainActor
final class RedirectUriOverridingIdpHandler: NSObject, @preconcurrency IdpHandler {
  /// Wrapped native handler for the selected provider.
  private var baseHandler: any IdpHandler

  /// Redirect URI override forwarded from the React Native bridge.
  private let redirectUri: String

  /// Creates a redirect-overriding handler wrapper.
  /// - Parameters:
  ///   - baseHandler: Native provider handler resolved for the selected callback provider.
  ///   - redirectUri: Redirect URI to inject into the `IdpClient`.
  init(baseHandler: any IdpHandler, redirectUri: String) {
    self.baseHandler = baseHandler
    self.redirectUri = redirectUri
    super.init()
  }

  /// Token type exposed by the wrapped provider handler.
  var tokenType: String {
    get {
      baseHandler.tokenType
    }
    set {
      baseHandler.tokenType = newValue
    }
  }

  /// Authorizes the wrapped provider after overriding the `IdpClient.redirectUri`.
  /// - Parameter idpClient: Native client payload created from the Journey callback.
  /// - Returns: The provider authorization result.
  /// - Throws: Any authorization error emitted by the wrapped handler.
  func authorize(idpClient: IdpClient) async throws -> IdpResult {
    var overriddenClient = idpClient
    overriddenClient.redirectUri = redirectUri
    return try await baseHandler.authorize(idpClient: overriddenClient)
  }
}

/// Shared iOS implementation for the Ping External IdP React Native module.
public class RNPingExternalIdpCommon: NSObject {
  private static let loggerIdKey = "loggerId"
  private static let redirectUriKey = "redirectUri"

  /// Stable error codes emitted by the External IdP module.
  ///
  /// Keep these in sync with JS `ExternalIdpErrorCode` and Android `ExternalIdpErrorCodes`.
  private enum ExternalIdpErrorCode: String {
    case authorizeError = "EXTERNAL_IDP_AUTHORIZE_ERROR"
    case cancelled = "EXTERNAL_IDP_CANCELLED"
    case unsupportedProvider = "EXTERNAL_IDP_UNSUPPORTED_PROVIDER"
    case callbackNotFound = "EXTERNAL_IDP_CALLBACK_NOT_FOUND"
    case configError = "EXTERNAL_IDP_CONFIG_ERROR"
    case windowUnavailable = "EXTERNAL_IDP_WINDOW_UNAVAILABLE"
  }

  /// Launches the external IdP authorization flow for an active Journey `IdpCallback`.
  /// - Parameters:
  ///   - journeyId: Native Journey instance id.
  ///   - options: Per-call options payload.
  ///   - config: Per-call runtime configuration payload.
  ///   - resolver: Promise resolver for the authorization result.
  ///   - rejecter: Promise rejecter for errors.
  @objc
  @MainActor
  public static func authorizeForJourney(
    _ journeyId: String,
    options: NSDictionary,
    config: NSDictionary,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock
  ) {
    let callConfig = parseCallConfig(config)
    let handlers = PromiseBridge<NSDictionary>(resolver: resolver, rejecter: rejecter)
    if journeyId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
      handlers.reject(
        GenericError(
          type: .argumentError,
          error: ExternalIdpErrorCode.callbackNotFound.rawValue,
          message: "Journey id must not be empty for external IdP authorization."
        )
      )
      return
    }

    guard PresentationAnchorResolver.resolveForegroundWindowAnchor() != nil else {
      handlers.reject(
        GenericError(
          type: .authError,
          error: ExternalIdpErrorCode.windowUnavailable.rawValue,
          message: "No active application window is available for external IdP authorization."
        )
      )
      return
    }

    let callbackIndex = parseCallbackIndex(options)

    // `@objc` methods cannot be declared `async`, so a Task is required to
    // enter the async context. The `@MainActor` annotation is redundant here
    // but kept to make the actor affinity explicit at the call site.
    Task { @MainActor in
      let logger = await resolveLoggerFromCore(callConfig.loggerId)
      logger?.i("External IdP authorizeForJourney requested for callback index \(callbackIndex)")
      guard let callback = await resolveIdpCallback(journeyId: journeyId, index: callbackIndex) else {
        logger?.w("External IdP authorizeForJourney callback not found at index \(callbackIndex)", error: nil)
        handlers.reject(
          GenericError(
            type: .stateError,
            error: ExternalIdpErrorCode.callbackNotFound.rawValue,
            message: "No active IdP callback found for journey \(journeyId) at index \(callbackIndex)."
          )
        )
        return
      }

      let result = await authorizeForJourney(callback: callback, redirectUri: callConfig.redirectUri)
      switch result {
      case .success(let idpResult):
        logger?.d("External IdP authorizeForJourney succeeded")
        handlers.resolve(createAuthorizeResultPayload(idpResult))

      case .failure(let error):
        let mapped = mapIdpError(error)
        logger?.e("External IdP authorizeForJourney failed with code \(mapped.error)", error: error as NSError)
        handlers.reject(
          mapped,
          underlying: error as NSError
        )
      }
    }
  }

  /// Mutates the native `SelectIdpCallback` state for an active Journey.
  /// - Parameters:
  ///   - journeyId: Native Journey instance id.
  ///   - provider: Provider identifier selected by the user.
  ///   - options: Per-call options payload.
  ///   - config: Per-call runtime configuration payload.
  ///   - resolver: Promise resolver for completion.
  ///   - rejecter: Promise rejecter for errors.
  @objc
  @MainActor
  public static func selectProviderForJourney(
    _ journeyId: String,
    provider: String,
    options: NSDictionary,
    config: NSDictionary,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock
  ) {
    let callConfig = parseCallConfig(config)
    let handlers = PromiseBridge<NSNull>(resolver: resolver, rejecter: rejecter)
    if journeyId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
      handlers.reject(
        GenericError(
          type: .argumentError,
          error: ExternalIdpErrorCode.callbackNotFound.rawValue,
          message: "Journey id must not be empty for setting selected provider."
        )
      )
      return
    }

    let selectedProvider = provider.trimmingCharacters(in: .whitespacesAndNewlines)
    if selectedProvider.isEmpty {
      handlers.reject(
        GenericError(
          type: .argumentError,
          error: ExternalIdpErrorCode.configError.rawValue,
          message: "Provider must not be empty for external IdP selection."
        )
      )
      return
    }

    let callbackIndex = parseCallbackIndex(options)

    // `@objc` methods cannot be declared `async`, so a Task is required to
    // enter the async context. The `@MainActor` annotation is redundant here
    // but kept to make the actor affinity explicit at the call site.
    Task { @MainActor in
      let logger = await resolveLoggerFromCore(callConfig.loggerId)
      logger?.i("External IdP select provider requested for callback index \(callbackIndex)")
      guard let callback = await resolveSelectIdpCallback(journeyId: journeyId, index: callbackIndex) else {
        logger?.w("External IdP select provider callback not found at index \(callbackIndex)", error: nil)
        handlers.reject(
          GenericError(
            type: .stateError,
            error: ExternalIdpErrorCode.callbackNotFound.rawValue,
            message: "No active SelectIdp callback found for journey \(journeyId) at index \(callbackIndex)."
          )
        )
        return
      }

      callback.value = selectedProvider
      logger?.d("External IdP select provider succeeded")
      handlers.resolve(NSNull())
    }
  }

  /// Launches the external IdP authorization flow for an active DaVinci `IdpCollector`.
  ///
  /// Architecturally different from Journey: the IdP token flows through
  /// `daVinci.next()` via the native `RequestInterceptor`. This method resolves
  /// void — the token is NOT returned to JS.
  ///
  /// - Parameters:
  ///   - davinciId: Native DaVinci instance id.
  ///   - options: Per-call options payload.
  ///   - config: Per-call runtime configuration payload.
  ///   - resolver: Promise resolver for void completion.
  ///   - rejecter: Promise rejecter for errors.
  @objc
  @MainActor
  public static func authorizeForDaVinci(
    _ davinciId: String,
    options: NSDictionary,
    config: NSDictionary,
    resolver: @escaping RCTPromiseResolveBlock,
    rejecter: @escaping RCTPromiseRejectBlock
  ) {
    let callConfig = parseCallConfig(config)
    let handlers = PromiseBridge<NSNull>(resolver: resolver, rejecter: rejecter)
    if davinciId.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
      handlers.reject(
        GenericError(
          type: .argumentError,
          error: ExternalIdpErrorCode.callbackNotFound.rawValue,
          message: "DaVinci id must not be empty for external IdP authorization."
        )
      )
      return
    }

    guard PresentationAnchorResolver.resolveForegroundWindowAnchor() != nil else {
      handlers.reject(
        GenericError(
          type: .authError,
          error: ExternalIdpErrorCode.windowUnavailable.rawValue,
          message: "No active application window is available for external IdP authorization."
        )
      )
      return
    }

    let collectorIndex = parseCallbackIndex(options)

    Task { @MainActor in
      let logger = await resolveLoggerFromCore(callConfig.loggerId)
      logger?.i("External IdP authorizeForDaVinci requested for collector index \(collectorIndex)")

      guard let collectors = await CoreRuntime.resolveDaVinciCollectors(davinciId) else {
        logger?.w("External IdP authorizeForDaVinci collectors not found for davinciId \(davinciId)", error: nil)
        handlers.reject(
          GenericError(
            type: .stateError,
            error: ExternalIdpErrorCode.callbackNotFound.rawValue,
            message: "No active DaVinci collectors found for id \(davinciId)."
          )
        )
        return
      }

      let matching = collectors.compactMap { $0 as? IdpCollector }
      guard collectorIndex >= 0, collectorIndex < matching.count else {
        logger?.w("External IdP authorizeForDaVinci collector not found at index \(collectorIndex)", error: nil)
        handlers.reject(
          GenericError(
            type: .stateError,
            error: ExternalIdpErrorCode.callbackNotFound.rawValue,
            message: "No active IdP collector found for DaVinci \(davinciId) at index \(collectorIndex)."
          )
        )
        return
      }

      let collector = matching[collectorIndex]
      let callbackURLScheme = callConfig.redirectUri.isEmpty ? nil : callConfig.redirectUri
      let result = await collector.authorize(callbackURLScheme: callbackURLScheme)
      switch result {
      case .success:
        logger?.d("External IdP authorizeForDaVinci succeeded")
        handlers.resolve(NSNull())

      case .failure(let error):
        let mapped = mapIdpError(error)
        logger?.e("External IdP authorizeForDaVinci failed with code \(mapped.error)", error: error as NSError)
        handlers.reject(mapped, underlying: error as NSError)
      }
    }
  }

  /// Resolves a native logger from the shared Core logger registry.
  ///
  /// - Parameter loggerId: Logger handle identifier from JS.
  /// - Returns: Native logger instance, or `nil` when missing/invalid.
  private static func resolveLoggerFromCore(_ loggerId: String?) async -> PingLogger.Logger? {
    guard let loggerId, !loggerId.isEmpty else { return nil }
    guard let handle = await CoreRuntime.loggerRegistry.resolve(loggerId) as? LoggerHandleContract else {
      return nil
    }
    return handle.nativeLogger as? PingLogger.Logger
  }

  /// Parses per-call bridge configuration from the React Native payload.
  /// - Parameter config: Raw bridge configuration dictionary.
  /// - Returns: Trimmed call configuration values.
  static func parseCallConfig(_ config: NSDictionary) -> ExternalIdpCallConfig {
    let loggerId = (config[loggerIdKey] as? String)?
      .trimmingCharacters(in: .whitespacesAndNewlines)
      .nilIfEmpty
    let redirectUri = (config[redirectUriKey] as? String)?
      .trimmingCharacters(in: .whitespacesAndNewlines) ?? ""
    return ExternalIdpCallConfig(loggerId: loggerId, redirectUri: redirectUri)
  }

  /// Authorizes a Journey IdP callback, optionally overriding its native redirect URI.
  /// - Parameters:
  ///   - callback: Journey-scoped callback resolved from Core runtime.
  ///   - redirectUri: Optional redirect URI override forwarded from JS.
  /// - Returns: Native authorization result.
  // TODO: Add browser fallback here when the iOS native SDK supports it for AIC Journey flows.
  @MainActor
  static func authorizeForJourney(
    callback: IdpCallback,
    redirectUri: String
  ) async -> Result<IdpResult, IdpExceptions> {
    guard !redirectUri.isEmpty else {
      return await callback.authorize()
    }

    guard let handler = resolveIdpHandler(for: callback.provider) else {
      return .failure(.unsupportedIdpException(message: callback.provider))
    }

    let overridingHandler = RedirectUriOverridingIdpHandler(
      baseHandler: handler,
      redirectUri: redirectUri
    )
    return await callback.authorize(idpHandler: overridingHandler)
  }

  /// Resolves one `IdpCallback` from active Journey callbacks by per-type index.
  ///
  /// - Parameters:
  ///   - journeyId: Native Journey instance id.
  ///   - index: Per-type callback index.
  /// - Returns: Matching callback when available.
  private static func resolveIdpCallback(journeyId: String, index: Int) async -> IdpCallback? {
    guard let callbacks = await CoreRuntime.resolveJourneyCallbacks(journeyId) else {
      return nil
    }
    let matching = callbacks.compactMap { $0 as? IdpCallback }
    guard index >= 0, index < matching.count else {
      return nil
    }
    return matching[index]
  }

  /// Resolves one `SelectIdpCallback` from active Journey callbacks by per-type index.
  ///
  /// - Parameters:
  ///   - journeyId: Native Journey instance id.
  ///   - index: Per-type callback index.
  /// - Returns: Matching callback when available.
  private static func resolveSelectIdpCallback(
    journeyId: String,
    index: Int
  ) async -> SelectIdpCallback? {
    guard let callbacks = await CoreRuntime.resolveJourneyCallbacks(journeyId) else {
      return nil
    }
    let matching = callbacks.compactMap { $0 as? SelectIdpCallback }
    guard index >= 0, index < matching.count else {
      return nil
    }
    return matching[index]
  }

  /// Parses callback index from callback execution options.
  ///
  /// - Parameter options: Callback execution options payload.
  /// - Returns: Parsed callback index, defaulting to 0.
  static func parseCallbackIndex(_ options: NSDictionary) -> Int {
    if let value = options["index"] as? NSNumber {
      return value.intValue
    }
    if let value = options["index"] as? String, let parsed = Int(value) {
      return parsed
    }
    return 0
  }

  /// Resolves the default native handler for a Journey `IdpCallback` provider.
  /// - Parameter provider: Provider identifier configured on the callback.
  /// - Returns: Matching native provider handler when available.
  @MainActor
  static func resolveIdpHandler(for provider: String) -> (any IdpHandler)? {
    let lowercasedProvider = provider.lowercased()

    switch true {
    case lowercasedProvider.contains("apple") || lowercasedProvider.contains("siwa"):
      return makeIdpHandler(from: "PingExternalIdPApple.AppleHandler")

    case lowercasedProvider.contains("google"):
      return makeIdpHandler(from: "PingExternalIdPGoogle.GoogleHandler")

    case lowercasedProvider.contains("facebook"):
      return makeIdpHandler(from: "PingExternalIdPFacebook.FacebookHandler")

    default:
      return nil
    }
  }

  /// Instantiates a native provider handler from its Objective-C runtime name.
  /// - Parameter className: Objective-C runtime class name for the provider handler.
  /// - Returns: Initialized handler when the provider SDK is linked.
  @MainActor
  private static func makeIdpHandler(from className: String) -> (any IdpHandler)? {
    guard let handlerClass = NSClassFromString(className) as? (NSObject & IdpHandler).Type else {
      return nil
    }
    return handlerClass.init()
  }

  /// Creates the JS-facing authorize result payload.
  ///
  /// - Parameter result: Native IdP authorization result.
  /// - Returns: Bridge-safe result dictionary.
  static func createAuthorizeResultPayload(_ result: IdpResult) -> NSDictionary {
    var payload: [String: Any] = ["token": result.token]
    if let additionalParameters = result.additionalParameters {
      payload["additionalParameters"] = additionalParameters
    }
    return payload as NSDictionary
  }

  /// Maps native IdP failures to the shared JS error contract.
  ///
  /// - Parameter error: Native IdP exception.
  /// - Returns: Generic bridge error.
  static func mapIdpError(_ error: IdpExceptions) -> GenericError {
    switch error {
    case .idpCanceledException:
      return GenericError(
        type: .authError,
        error: ExternalIdpErrorCode.cancelled.rawValue,
        message: error.localizedDescription
      )

    case .unsupportedIdpException:
      return GenericError(
        type: .authError,
        error: ExternalIdpErrorCode.unsupportedProvider.rawValue,
        message: error.localizedDescription
      )

    case .illegalArgumentException:
      return GenericError(
        type: .argumentError,
        error: ExternalIdpErrorCode.configError.rawValue,
        message: error.localizedDescription
      )

    case .illegalStateException:
      return GenericError(
        type: .authError,
        error: ExternalIdpErrorCode.authorizeError.rawValue,
        message: error.localizedDescription
      )
    }
  }
}

private extension String {
  /// Returns `nil` when the string becomes empty after trimming.
  var nilIfEmpty: String? {
    isEmpty ? nil : self
  }
}
